qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,632 @@
+                raise EvalProxyError("Invalid number of args")
+            nargs = int(len(args) / 2)
+
+            for e, i in zip(order[:nargs], itertools.count(0)):
+                args1[e] = args[i]
+                args2[e] = args[nargs + i]
+
+        if not args1:
+            raise EvalProxyError("Nothing is passed to check the range")
+
+        return self._time_range(args1, args2, gmt)
+
+
+class PACResolver(object):
+    """Evaluate PAC script files and resolve proxies."""
+
+    @staticmethod
+    def _parse_proxy_host(host_str):
+        host, _colon, port_str = host_str.partition(':')
+        try:
+            port = int(port_str)
+        except ValueError:
+            raise ParseProxyError("Invalid port number")
+        return (host, port)
+
+    @staticmethod
+    def _parse_proxy_entry(proxy_str):
+        """Parse one proxy string entry, as described in PAC specification."""
+        config = [c for c in proxy_str.split(' ') if c]
+        if not config:
+            raise ParseProxyError("Empty proxy entry")
+        elif config[0] == "DIRECT":
+            if len(config) != 1:
+                raise ParseProxyError("Invalid number of parameters")
+            return QNetworkProxy(QNetworkProxy.NoProxy)
+        elif config[0] == "PROXY":
+            if len(config) != 2:
+                raise ParseProxyError("Invalid number of parameters")
+            host, port = PACResolver._parse_proxy_host(config[1])
+            return QNetworkProxy(QNetworkProxy.HttpProxy, host, port)
+        elif config[0] == "SOCKS":
+            if len(config) != 2:
+                raise ParseProxyError("Invalid number of parameters")
+            host, port = PACResolver._parse_proxy_host(config[1])
+            return QNetworkProxy(QNetworkProxy.Socks5Proxy, host, port)
+        else:
+            err = "Unknown proxy type: {}"
+            raise ParseProxyError(err.format(config[0]))
+
+    @staticmethod
+    def _parse_proxy_string(proxy_str):
+        proxies = proxy_str.split(';')
+        return [PACResolver._parse_proxy_entry(x) for x in proxies]
+
+    def __init__(self, pac_str, fixed_dt=None):
+        """Create a PAC resolver.
+
+        Args:
+            pac_str: JavaScript code containing PAC resolver.
+            fixed_dt: Assume fixed date and time instead of system time.
+        """
+        self._engine = QJSEngine()
+
+        self._ctx = _PACContext(self._engine, fixed_dt=fixed_dt)
+        self._engine.globalObject().setProperty("PAC",
+            self._engine.newQObject(self._ctx))
+        proxy_config = self._engine.newObject()
+        proxy_config.setProperty("bindings", self._engine.newObject())
+        self._engine.globalObject().setProperty("ProxyConfig", proxy_config)
+        ctx_meta = self._ctx.metaObject()
+        for i in range(ctx_meta.methodCount()):
+            m = ctx_meta.method(i)
+            if m.typeName() == "QJSValue":
+                name = bytes(m.name()).decode()
```

Comment:
Ah! I thought it would use the system default encoding by default like reading a file does, but looks like it defaults to UTF-8. I'd still prefer explicitly passing `utf-8` here as IMHO what happens without an encoding given is quite confusing.